Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -52,8 +53,18 @@ void WebPermissionController::query(WebCore::ClientOrigin&& origin, WebCore::Per
     if (cachedResult != WebCore::PermissionState::Prompt)
         return completionHandler(cachedResult);
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
-    tryProcessingRequests();
+    if (async == WebCore::Asynchronous::Yes) {
+        m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
+        tryProcessingRequests();
+    } else {
+        std::optional<WebCore::PermissionState> state;
+        bool shouldCache;
+        m_page->sendSync(Messages::WebPageProxy::QueryPermissionSync(origin, descriptor), Messages::WebPageProxy::QueryPermissionSync::Reply(state, shouldCache));
+        
+        completionHandler(state);
+        if (shouldCache && state)
```

Comment:
IIUC, the result is never cached and we will send a sync message to UI process per request. I think this is not ideal and we can have better implementation after we support permission change from UI process, i.e. caching the result and let UI process push updates; we may even eliminate the need for sync message and just return default when cache is empty (e.g. if page has not called query/requestPermission). 
So it's probably a better idea to implement change event first.

Reply:
I think the sync IPC is acceptable for now, in order to have a single code path for both the sync and async APIs. Otherwise, some of Rupin's work ends up getting duplicated.

Reply:
I would expect the duplicate work to be little after what Rupin has already done (like the permission requester cache), at least in Worker support and change event support, and we would have more flexibility in implementation (e.g. using async message instead of sync message) with these support, so I think it's better getting those two items done first.

Reply:
Ok, I think Rupin wanted to do this patch because he wanted to move his requester cache from the WebPageProxy to the WebsiteDataStore. Right now the cache used by the old code path in the WebProcess would cause inconsistency between the sync and async APIs.

Anyway, I guess we can keep the requester cache on the WebPageProxy for now to match the WebContent process side requester cache that is used by the legacy sync API.